### PR TITLE
Add versions of the route-adding methods that include a "validator" block.

### PR DESCRIPTION
**Pull request by vokal-isaac**
If present, the validator block is run during route-matching, if a route appears to match based on the pattern.  If the validator returns YES, the route is considered a match; if the validator returns NO, the route is not considered a match.

@vokal/ios-developers Code review please?

### CODE REVIEW

File: JLRoutes/JLRoutes.h [hunk 1]
```diff
@@ -38,6 +38,12 @@ static NSString *const kJLRoutesGlobalNamespaceKey = @"JLRoutesGlobalNamespace";
 /// Registers a routePattern with default priority (0) in the receiving scheme namespace.
 + (void)addRoute:(NSString *)routePattern handler:(BOOL (^)(NSDictionary *parameters))handlerBlock;
 - (void)addRoute:(NSString *)routePattern handler:(BOOL (^)(NSDictionary *parameters))handlerBlock; // instance method
++ (void)addRoute:(NSString *)routePattern
+       validator:(BOOL (^)(NSDictionary *parameters))validatorBlock
```
Comment (chillpop) [inline JLRoutes/JLRoutes.h]: any interest in `typedef`ing this block signature? (or is that outside the scope of this PR?)

**Comment by brockboland** [comment]
Are you going to submit a PR upstream on the original repo?

File: JLRoutes/JLRoutes.m [hunk 2]
```diff
@@ -192,8 +193,22 @@ + (void)addRoute:(NSString *)routePattern handler:(BOOL (^)(NSDictionary *parame
 }
 
 
++ (void)addRoute:(NSString *)routePattern validator:(BOOL (^)(NSDictionary *))validatorBlock handler:(BOOL (^)(NSDictionary *))handlerBlock {
+    [[self globalRoutes] addRoute:routePattern validator:validatorBlock handler:handlerBlock];
+}
+
+
```
Comment (chillpop) [inline JLRoutes/JLRoutes.m:200]: is two blank lines between methods the standard here?

**Comment by vokal-isaac** [comment]
My plan was to submit upstream, yes.

File: JLRoutes/JLRoutes.h (hunk 1 above)
Comment (vokal-isaac) [inline JLRoutes/JLRoutes.h]: It'd probably make sense to do it... The only reason I hadn't was that it already existed that way, which isn't a good reason.

File: JLRoutes/JLRoutes.m (hunk 2 above)
Comment (vokal-isaac) [inline JLRoutes/JLRoutes.m:200]: Seems to be.

File: JLRoutes/JLRoutes.m [hunk 3]
```diff
@@ -202,18 +217,37 @@ - (void)addRoute:(NSString *)routePattern handler:(BOOL (^)(NSDictionary *parame
 }
 
 
+- (void)addRoute:(NSString *)routePattern validator:(BOOL (^)(NSDictionary *))validatorBlock handler:(BOOL (^)(NSDictionary *))handlerBlock {
+    [self addRoute:routePattern priority:0 validator:validatorBlock handler:handlerBlock];
+}
+
```
Comment (chillpop) [inline JLRoutes/JLRoutes.m:223]: if 2 blank lines is standard, you're missing a blank line here

File: JLRoutes/JLRoutes.m [hunk 4]
```diff
@@ -202,18 +217,37 @@ - (void)addRoute:(NSString *)routePattern handler:(BOOL (^)(NSDictionary *parame
...
+        priority:(NSUInteger)priority
+       validator:(BOOL (^)(NSDictionary *))validatorBlock
+         handler:(BOOL (^)(NSDictionary *))handlerBlock
+{
 	_JLRoute *route = [[_JLRoute alloc] init];
 	route.pattern = routePattern;
 	route.priority = priority;
-	route.block = [handlerBlock copy];
+    route.validatorBlock = [validatorBlock copy];
+    route.helperBlock = [handlerBlock copy];
```
Comment (chillpop) [inline JLRoutes/JLRoutes.m]: indentation is off here: looks like they use tabs, not spaces

File: JLRoutes/JLRoutes.m [hunk 5]
```diff
@@ -69,7 +69,8 @@ @interface _JLRoute : NSObject
 
 @property (nonatomic, weak) JLRoutes *parentRoutesController;
 @property (nonatomic, strong) NSString *pattern;
-@property (nonatomic, strong) BOOL (^block)(NSDictionary *parameters);
+@property (nonatomic, strong) BOOL (^validatorBlock)(NSDictionary *parameters);
+@property (nonatomic, strong) BOOL (^helperBlock)(NSDictionary *parameters);
```
Comment (chillpop) [inline JLRoutes/JLRoutes.m]: if you declare these `copy` can you eliminate the copy calls below?

**Comment by brockboland** [comment]
:+1:

File: JLRoutes/JLRoutes.m (hunk 5 above)
Comment (vokal-isaac) [inline JLRoutes/JLRoutes.m]: I believe so.

**Comment by vokal-isaac** [comment]
I think I got all the spaces converted to tabs (some of them weren't mine).

**Comment by chillpop** [comment]
LGTM

**Comment by seanwolter** [comment]
:+1:

**Comment by brockboland** [comment]
LGTM2

**Comment by brockboland** [comment]
oh wait nevermind no one cares anymore

**Comment by seanwolter** [comment]
^^ fav

**Pull request merged**